File: Python/AlgoBootcamp/leetcode/medium/1466. Reorder Routes to Make All Paths Lead to the City Zero.py
# There are n cities numbered from 0 to n - 1 and n - 1 roads such that there is only one way to travel between two different cities (this network form a tree). Last year, The ministry of transport decided to orient the roads in one direction because they are too narrow.
#
# Roads are represented by connections where connections[i] = [ai, bi] represents a road from city ai to city bi.
#
# This year, there will be a big event in the capital (city 0), and many people want to travel to this city.
#
# Your task consists of reorienting some roads such that each city can visit the city 0. Return the minimum number of edges changed.
#
# It's guaranteed that each city can reach city 0 after reorder.
#
#
#
# Example 1:
#
#
# Input: n = 6, connections = [[0,1],[1,3],[2,3],[4,0],[4,5]]
# Output: 3
# Explanation: Change the direction of edges show in red such that each node can reach the node 0 (capital).
# Example 2:
#
#
# Input: n = 5, connections = [[1,0],[1,2],[3,2],[3,4]]
# Output: 2
# Explanation: Change the direction of edges show in red such that each node can reach the node 0 (capital).
# Example 3:
#
# Input: n = 3, connections = [[1,0],[2,0]]
# Output: 0
#
#
# Constraints:
#
# 2 <= n <= 5 * 104
# connections.length == n - 1
# connections[i].length == 2
# 0 <= ai, bi <= n - 1
# ai != bi
from collections import defaultdict
from typing import List


class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        adj = defaultdict(list) # original direction
        dirrections = set()
        for a, b in connections:
            adj[a].append(b)
            adj[b].append(a)
            dirrections.add((a, b))
        visited = {0}
        result = 0

        def dfs(node):
            nonlocal result
            visited.add(node)
            for nei in adj[node]:
                if nei in visited:
                    continue
                # wrong direction, should be reversed
                if (node, nei) in dirrections:
                    result += 1
                dfs(nei)

        dfs(0)
        return result


sol = Solution()
res = sol.minReorder(2, [[0, 1]])
assert res == 1

File: Kotlin/src/main/kotlin/leetcode/medium/1466 Reorder Routes to Make All Paths Lead to the City Zero.kt
package leetcode.medium

class `1466 Reorder Routes to Make All Paths Lead to the City Zero` {
    fun minReorder(n: Int, connections: Array<IntArray>): Int {
        val adj = mutableMapOf<Int, MutableList<Int>>() // original direction
        val directions = mutableSetOf<Pair<Int, Int>>()
        for (connection in connections) {
            adj.getOrPut(connection[0]) { mutableListOf() }.add(connection[1])
            adj.getOrPut(connection[1]) { mutableListOf() }.add(connection[0])
            directions.add(Pair(connection[0], connection[1]))
        }
        val visited = mutableSetOf(0)
        var result = 0

        fun dfs(node: Int) {
            visited.add(node)
            for (nei in adj.getOrDefault(node, mutableListOf())) {
                if (nei in visited) {
                    continue
                }
                // wrong direction, should be reversed
                if (Pair(node, nei) in directions) {
                    result++
                }
                dfs(nei)
            }
        }

        dfs(0)
        return result
    }
}

fun main() {
    val solution = `1466 Reorder Routes to Make All Paths Lead to the City Zero`()
    val connections1 = arrayOf(
        intArrayOf(0, 1),
        intArrayOf(1, 3),
        intArrayOf(2, 3),
        intArrayOf(4, 0),
        intArrayOf(4, 5)
    )
    assert(solution.minReorder(6, connections1) == 3)

    val connections2 = arrayOf(
        intArrayOf(1, 0),
        intArrayOf(2, 0)
    )
    assert(solution.minReorder(3, connections2) == 0)
}